web/packages/shared/components/FieldSelect/FieldSelect.tsx

```diff
@@ -84,9 +87,84 @@ export default function FieldSelect({
   );
 }
 
-type Props = SelectProps & {
+/** @deprecated Use the named export `{ FieldSelect }`. */
+export default FieldSelect;
+
+export function FieldSelectAsync({
+  components,
+  label,
+  labelTip,
+  value,
+  name,
+  onChange,
+  placeholder,
+  maxMenuHeight,
+  isClearable,
+  isMulti,
+  menuPosition,
+  rule = defaultRule,
+  stylesConfig,
+  isSearchable,
+  isSimpleValue,
+  autoFocus,
+  isDisabled,
+  elevated,
+  noOptionsMessage,
+  loadOptions,
+  inputId = 'select',
+  ...styles
+}: AsyncSelectProps & FieldProps) {
+  const [attempt, runAttempt] = useAsync(loadOptions);
```

Comment:
Did you consider passing `attempt` and `runAttempt` from outside?

I think it could make the component easier to use, as then you would be able to separate how data is fetched from how it is displayed. E.g. you could have `FieldSelectAsync` which fetches data when the field is active, but you could also have a variant of it that prefetches data, without having to modify the source of `FieldSelectAsync`.

In that sense, `FieldSelectAsync` would boil down to knowing how to handle `Attempt`.

Reply:
I didn't consider it, but I think it won't work here.
The only way to give options to the select component is to return them from `loadOptions` callback. 
Unfortunately, something like the following won't work because the returned options would be "off by one" (at the moment of invoking the function, `attempt.data` would still contain the previous value):
```ts
        loadOptions={async (input, option) => {
          await runAttempt(input, option);
          return attempt.status === 'success' && attempt.data;
        }}
```
This means that wouldn't be able to fetch the options outside and pass them to the component.

However, prefetching is done by the select component itself, I passed the option `defaultOptions: true` which does this for us.

Reply:
out of curiosity for if i use this, how would you re-attempt, if there was an error? (if user could not refresh the step they are in)?

Reply:
> The only way to give options to the select component is to return them from `loadOptions` callback.

Ah, I see. I asked because I know that we have a mix of components, some that let you specify just data to display and some that ask you to specify _how_ to fetch data. Here I understand it's dictated by React Select's API. But I remember I added `useAsync` directly to `MenuLogin` and now I think that was a mistake.

Reply:
>Here I understand it's dictated by React Select's API. 

Yeah. The only way to overcome this would be to use a regular select component and supply our own logic for fetching data: prefetching in `useEffect`, manually firing requests when the input value changes. But I'm not sure if this is worth the effort, I'm leaning towards using the existing `AsyncSelect`. 

>Re re-attempt, maybe we could document this then?

👍